File: java_client/examples/BasicExample.java
/**
 * BasicExample - Basic FastDFS Client Usage Example
 * 
 * This example demonstrates basic usage of the FastDFS Java client,
 * including file upload, download, and deletion operations.
 * 
 * Copyright (C) 2025 FastDFS Java Client Contributors
 * 
 * FastDFS may be copied only under the terms of the GNU General
 * Public License V3, which may be found in the FastDFS source kit.
 * 
 * @author FastDFS Java Client Contributors
 * @version 1.0.0
 * @since 1.0.0
 */
package com.fastdfs.client.examples;

import com.fastdfs.client.FastDFSClient;
import com.fastdfs.client.config.FastDFSConfig;
import com.fastdfs.client.exception.FastDFSException;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * BasicExample demonstrates basic FastDFS operations.
 * 
 * This example shows how to:
 * - Create a FastDFS client with configuration
 * - Upload a file from the local filesystem
 * - Upload file data from a byte array
 * - Download a file
 * - Delete a file
 * - Handle errors
 * 
 * To run this example:
 * 1. Make sure FastDFS tracker and storage servers are running
 * 2. Update the tracker server addresses in the configuration
 * 3. Compile and run this class
 */
public class BasicExample {
    
    /**
     * Main method - Entry point for the example.
     * 
     * @param args command line arguments (not used)
     */
    public static void main(String[] args) {
        // Create client configuration
        // Replace with your actual tracker server addresses
        FastDFSConfig config = new FastDFSConfig.Builder()
            .addTrackerServer("192.168.1.100", 22122)
            .addTrackerServer("192.168.1.101", 22122)  // Optional: add more for high availability
            .maxConnectionsPerServer(100)                // Maximum connections per server
            .connectTimeout(5000)                        // Connection timeout: 5 seconds
            .networkTimeout(30000)                       // Network I/O timeout: 30 seconds
            .idleTimeout(60000)                          // Idle connection timeout: 60 seconds
            .retryCount(3)                               // Retry failed operations up to 3 times
            .build();
        
        // Initialize client
        // The client should be created once and reused throughout the application
        FastDFSClient client = null;
        try {
            client = new FastDFSClient(config);
            
            // Example 1: Upload a file from the local filesystem
            System.out.println("Example 1: Upload file from filesystem");
            String fileId1 = uploadFileExample(client);
            System.out.println("File uploaded with ID: " + fileId1);
            
            // Example 2: Upload file data from a byte array
            System.out.println("\nExample 2: Upload file from byte array");
            String fileId2 = uploadBufferExample(client);
            System.out.println("File uploaded with ID: " + fileId2);
            
            // Example 3: Download a file
            System.out.println("\nExample 3: Download file");
            downloadFileExample(client, fileId2);
            
            // Example 4: Delete a file
            System.out.println("\nExample 4: Delete file");
            deleteFileExample(client, fileId1);
            deleteFileExample(client, fileId2);
            
            System.out.println("\nAll examples completed successfully!");
            
        } catch (FastDFSException e) {
            // Handle FastDFS-specific errors
            System.err.println("FastDFS error: " + e.getError());
            System.err.println("Message: " + e.getMessage());
            if (e.getCause() != null) {
                System.err.println("Cause: " + e.getCause().getMessage());
            }
            e.printStackTrace();
        } catch (IOException e) {
            // Handle I/O errors
            System.err.println("I/O error: " + e.getMessage());
            e.printStackTrace();
        } catch (Exception e) {
            // Handle other errors
            System.err.println("Unexpected error: " + e.getMessage());
            e.printStackTrace();
        } finally {
            // Always close the client to release resources
            if (client != null) {
                try {
                    client.close();
                    System.out.println("\nClient closed successfully");
                } catch (FastDFSException e) {
                    System.err.println("Error closing client: " + e.getMessage());
                }
            }
        }
    }
    
    /**
     * Example: Upload a file from the local filesystem.
     * 
     * @param client the FastDFS client
     * @return the file ID of the uploaded file
     * @throws FastDFSException if the upload fails
     * @throws IOException if the file cannot be read
     */
    private static String uploadFileExample(FastDFSClient client) 
            throws FastDFSException, IOException {
        // Upload a file from the local filesystem
        // The file extension is automatically extracted from the filename
        String localFilePath = "test.jpg";  // Replace with your file path
        
        // Optional: Add metadata to the file
        Map<String, String> metadata = new HashMap<>();
        metadata.put("author", "John Doe");
        metadata.put("description", "Test image file");
        metadata.put("uploaded-by", "BasicExample");
        
        // Upload the file
        String fileId = client.uploadFile(localFilePath, metadata);
        
        return fileId;
    }
    
    /**
     * Example: Upload file data from a byte array.
     * 
     * @param client the FastDFS client
     * @return the file ID of the uploaded file
     * @throws FastDFSException if the upload fails
     */
    private static String uploadBufferExample(FastDFSClient client) 
            throws FastDFSException {
        // Create some test data
        String content = "Hello, FastDFS! This is a test file.";
        byte[] data = content.getBytes();
        
        // Upload the data
        // The file extension must be provided explicitly
        String fileId = client.uploadBuffer(data, "txt", null);
        
        return fileId;
    }
    
    /**
     * Example: Download a file.
     * 
     * @param client the FastDFS client
     * @param fileId the file ID to download
     * @throws FastDFSException if the download fails
     */
    private static void downloadFileExample(FastDFSClient client, String fileId) 
            throws FastDFSException {
        // Download the entire file
        byte[] data = client.downloadFile(fileId);
        
        System.out.println("Downloaded " + data.length + " bytes");
        System.out.println("Content: " + new String(data));
        
        // Alternatively, download to a local file
        // client.downloadToFile(fileId, "downloaded_file.txt");
    }
    
    /**
     * Example: Delete a file.
     * 
     * @param client the FastDFS client
     * @param fileId the file ID to delete
     * @throws FastDFSException if the deletion fails
     */
    private static void deleteFileExample(FastDFSClient client, String fileId) 
            throws FastDFSException {
        // Delete the file
        client.deleteFile(fileId);
        System.out.println("File deleted: " + fileId);
    }
}

File: csharp_client/examples/BasicExample.cs
// ============================================================================
// FastDFS C# Client - Basic Example
// ============================================================================
// 
// Copyright (C) 2025 FastDFS C# Client Contributors
//
// This example demonstrates basic FastDFS operations including file upload,
// download, and deletion. It shows how to initialize the client, perform
// simple file operations, and handle errors.
//
// ============================================================================

using System;
using System.IO;
using System.Threading.Tasks;
using FastDFS.Client;

namespace FastDFS.Client.Examples
{
    /// <summary>
    /// Basic example demonstrating FastDFS file operations.
    /// 
    /// This example shows:
    /// - How to configure and initialize the FastDFS client
    /// - How to upload files to FastDFS storage
    /// - How to download files from FastDFS storage
    /// - How to delete files from FastDFS storage
    /// - How to handle errors and exceptions
    /// </summary>
    class BasicExample
    {
        /// <summary>
        /// Main entry point for the basic example.
        /// </summary>
        /// <param name="args">
        /// Command-line arguments (not used in this example).
        /// </param>
        /// <returns>
        /// A task that represents the asynchronous operation.
        /// </returns>
        static async Task Main(string[] args)
        {
            Console.WriteLine("FastDFS C# Client - Basic Example");
            Console.WriteLine("==================================");
            Console.WriteLine();

            // Step 1: Create client configuration
            // The configuration specifies tracker server addresses, timeouts,
            // connection pool settings, and other operational parameters.
            var config = new FastDFSClientConfig
            {
                // Specify tracker server addresses
                // Tracker servers coordinate file storage and retrieval operations
                TrackerAddresses = new[]
                {
                    "192.168.1.100:22122",  // Primary tracker server
                    "192.168.1.101:22122"   // Secondary tracker server (for redundancy)
                },

                // Maximum number of connections per server
                // Higher values allow more concurrent operations but consume more resources
                MaxConnections = 100,

                // Connection timeout: maximum time to wait when establishing connections
                ConnectTimeout = TimeSpan.FromSeconds(5),

                // Network timeout: maximum time for read/write operations
                NetworkTimeout = TimeSpan.FromSeconds(30),

                // Idle timeout: time before idle connections are closed
                IdleTimeout = TimeSpan.FromMinutes(5),

                // Retry count: number of retry attempts for failed operations
                RetryCount = 3
            };

            // Step 2: Initialize the FastDFS client
            // The client manages connections to tracker and storage servers,
            // handles connection pooling, and provides a high-level API for
            // file operations.
            using (var client = new FastDFSClient(config))
            {
                try
                {
                    // Step 3: Upload a file
                    // This example uploads a local file to FastDFS storage.
                    // The method returns a file ID that uniquely identifies
                    // the file in the FastDFS cluster.
                    Console.WriteLine("Step 1: Uploading file...");
                    var localFilePath = "test.txt";

                    // Create a test file if it doesn't exist
                    if (!File.Exists(localFilePath))
                    {
                        await File.WriteAllTextAsync(localFilePath, "Hello, FastDFS!");
                        Console.WriteLine($"Created test file: {localFilePath}");
                    }

                    // Upload the file
                    // The second parameter is metadata (null means no metadata)
                    var fileId = await client.UploadFileAsync(localFilePath, null);
                    Console.WriteLine($"File uploaded successfully!");
                    Console.WriteLine($"File ID: {fileId}");
                    Console.WriteLine();

                    // Step 4: Get file information
                    // Retrieve detailed information about the uploaded file,
                    // including size, creation time, CRC32 checksum, etc.
                    Console.WriteLine("Step 2: Getting file information...");
                    var fileInfo = await client.GetFileInfoAsync(fileId);
                    Console.WriteLine($"File Size: {fileInfo.FileSize} bytes");
                    Console.WriteLine($"Create Time: {fileInfo.CreateTime}");
                    Console.WriteLine($"CRC32: {fileInfo.CRC32:X8}");
                    Console.WriteLine($"Source IP: {fileInfo.SourceIPAddr}");
                    Console.WriteLine();

                    // Step 5: Download the file
                    // Download the file content as a byte array.
                    // For large files, consider using DownloadToFileAsync
                    // to stream directly to disk.
                    Console.WriteLine("Step 3: Downloading file...");
                    var downloadedData = await client.DownloadFileAsync(fileId);
                    var downloadedText = System.Text.Encoding.UTF8.GetString(downloadedData);
                    Console.WriteLine($"File downloaded successfully!");
                    Console.WriteLine($"Downloaded content: {downloadedText}");
                    Console.WriteLine($"Downloaded size: {downloadedData.Length} bytes");
                    Console.WriteLine();

                    // Step 6: Download to a local file
                    // This method streams the file directly to disk, which is
                    // more memory-efficient for large files.
                    Console.WriteLine("Step 4: Downloading file to disk...");
                    var downloadPath = "downloaded_test.txt";
                    await client.DownloadToFileAsync(fileId, downloadPath);
                    Console.WriteLine($"File downloaded to: {downloadPath}");
                    Console.WriteLine();

                    // Step 7: Download a partial file range
                    // Download only a specific byte range from the file.
                    // This is useful for large files where you only need
                    // a portion of the data.
                    Console.WriteLine("Step 5: Downloading file range...");
                    var rangeData = await client.DownloadFileRangeAsync(fileId, 0, 5);
                    var rangeText = System.Text.Encoding.UTF8.GetString(rangeData);
                    Console.WriteLine($"Downloaded range (0-5): {rangeText}");
                    Console.WriteLine();

                    // Step 8: Delete the file
                    // Permanently delete the file from FastDFS storage.
                    // This operation cannot be undone.
                    Console.WriteLine("Step 6: Deleting file...");
                    await client.DeleteFileAsync(fileId);
                    Console.WriteLine("File deleted successfully!");
                    Console.WriteLine();

                    // Clean up local files
                    if (File.Exists(localFilePath))
                    {
                        File.Delete(localFilePath);
                    }
                    if (File.Exists(downloadPath))
                    {
                        File.Delete(downloadPath);
                    }

                    Console.WriteLine("Example completed successfully!");
                }
                catch (FastDFSException ex)
                {
                    // Handle FastDFS-specific errors
                    Console.WriteLine($"FastDFS Error: {ex.Message}");
                    if (ex.InnerException != null)
                    {
                        Console.WriteLine($"Inner Exception: {ex.InnerException.Message}");
                    }
                }
                catch (Exception ex)
                {
                    // Handle other errors
                    Console.WriteLine($"Error: {ex.Message}");
                    Console.WriteLine($"Stack Trace: {ex.StackTrace}");
                }
            }

            Console.WriteLine();
            Console.WriteLine("Press any key to exit...");
            Console.ReadKey();
        }
    }
}

